plugins/modules/mso_schema_template_service_graph.py

```diff
@@ -0,0 +1,264 @@
+    service_node_list = module.params.get('service_node_list')
+    node_filter = module.params.get('node_filter')
+    state = module.params.get('state')
+
+    mso = MSOModule(module)
+
+    # Get schema
+    schema_id, schema_path, schema_obj = mso.query_schema(schema)
+
+    # Get template
+    templates = [t.get('name') for t in schema_obj.get('templates')]
+    if template not in templates:
+        mso.fail_json(msg="Provided template '{template}' does not exist. Existing templates: {templates}".format(template=template,
+                                                                                                                  templates=', '.join(templates)))
+    template_idx = templates.index(template)
+
+    mso.existing = {}
+    service_graph_idx = None
+    index = 0
+    Nodes = []
+
+    # Get service graphs
+    service_graphs = [f.get('name') for f in schema_obj.get('templates')[template_idx]['serviceGraphs']]
+    if service_graph_name in service_graphs:
+        service_graph_idx = service_graphs.index(service_graph_name)
+    
+    # Get service nodes
+    query_node_data = mso.query_nodes()
+    service_nodes = [f.get('name') for f in query_node_data]
+    if service_node_list is not None:
+      for node in service_node_list:
+        node_name = node.get('type')
+        if node_name in service_nodes:
+          index = index + 1
+          for node_data in query_node_data:
+            if node_data['name'] == node_name:
+              node_data = node_data
+          Nodes.append(dict(
+              name=node_name,
+              serviceNodeTypeId=node_data.get('id'),
+              index = index,
+              serviceNodeRef=dict(   
+                serviceNodeName=node_name,
+                serviceGraphName=service_graph_name,
+                templateName=template,
+                schemaId=schema_id,
+          )),
+          )
+  
+    if state == 'query':
+      if service_graph_name is None:
+            mso.existing = schema_obj.get('templates')[template_idx]['serviceGraphs']
+      if service_graph_name is not None and service_graph_idx is None:
+            mso.fail_json(msg="service_graph '{service_graph}' not found".format(service_graph=service_graph_name))
+      elif service_graph_idx is not None:
+            mso.existing = schema_obj.get('templates')[template_idx]['serviceGraphs'][service_graph_idx]
+      mso.exit_json()
+
+    service_graphs_path = '/templates/{0}/serviceGraphs'.format(template)
+    service_graph_path = '/templates/{0}/serviceGraphs/{1}'.format(template, service_graph_name)
+    service_nodes_path = '/templates/{0}/serviceGraphs/{1}/serviceNodes'.format(template, service_graph_name)
+    ops = []
+
+    mso.previous = mso.existing
+    if state == 'absent':
+        mso.proposed = mso.sent = {}
+        if service_graph_idx is None:
+            # There was no service graph to begin with
+            pass
+        else:
```

Comment:
I don't think this condition is needed as you don't use the idx in the API call.